components/script/dom/bindings/typedarrays.rs

```diff
@@ -0,0 +1,105 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#![allow(unsafe_code)]
+
+use std::ptr;
+
+use js::jsapi::{Heap, JSObject, JS_GetArrayBufferViewBuffer};
+use js::rust::wrappers::DetachArrayBuffer;
+use js::rust::{CustomAutoRooterGuard, MutableHandleObject};
+use js::typedarray::{CreateWith, Float32Array};
+
+use crate::script_runtime::JSContext;
+
+pub fn create_float32_array(
+    cx: JSContext,
+    data: &[f32],
+    dest: MutableHandleObject,
+) -> Result<Float32Array, ()> {
+    let res = unsafe { Float32Array::create(*cx, CreateWith::Slice(data), dest) };
+    if res.is_err() {
+        Err(())
+    } else {
+        Float32Array::from(dest.get())
+    }
+}
+
+#[derive(Default, JSTraceable)]
+pub struct HeapFloat32Array {
+    internal: Heap<*mut JSObject>,
```

Comment:
I just noticed that the existing audio buffer code is unsafe, because storing Heapin a Vec means the values can be moved in memory when the vec resizes. Let's make this a `Box<Heap<...>>` instead so it's impossible to misuse?

Reply:
https://doc.servo.org/mozjs_sys/jsgc/struct.Heap.html#method.boxed

Reply:
Good catch!